staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_interval.go

```diff
@@ -0,0 +1,207 @@
+	// is still valid or not.
+	valid bool
+}
+
+type attrFunc func(runtime.Object) (labels.Set, fields.Set, error)
+
+func newCacheInterval(startIndex, endIndex int, locker sync.Locker) *watchCacheInterval {
+	return &watchCacheInterval{
+		startIndex:     startIndex,
+		endIndex:       endIndex,
+		intervalBuffer: &watchCacheIntervalBuffer{buffer: make([]*watchCacheEvent, bufferSize)},
+		lock:           locker,
+	}
+}
+
+// newCacheIntervalFromStore is meant to handle the case of rv=0, such that the events
+// returned by Next() need to be events from a List() done on the underlying store of
+// the watch cache.
+func newCacheIntervalFromStore(store cache.Indexer, getAttrsFunc attrFunc, locker sync.Locker) (*watchCacheInterval, error) {
+	ciBuffer := &watchCacheIntervalBuffer{}
+	allItems := store.List()
+	ciBuffer.buffer = make([]*watchCacheEvent, len(allItems))
+	for i, item := range allItems {
+		elem, ok := item.(*storeElement)
+		if !ok {
+			return nil, fmt.Errorf("not a storeElement: %v", elem)
+		}
+		objLabels, objFields, err := getAttrsFunc(elem.Object)
+		if err != nil {
+			return nil, err
+		}
+		ciBuffer.buffer[i] = &watchCacheEvent{
+			Type:            watch.Added,
+			Object:          elem.Object,
+			ObjLabels:       objLabels,
+			ObjFields:       objFields,
+			Key:             elem.Key,
+			ResourceVersion: 0,
+		}
+	}
+	ci := &watchCacheInterval{intervalBuffer: ciBuffer, lock: locker}
+	// simulate that we already have all the events we're looking for.
+	ci.startIndex = ci.endIndex
+
+	return ci, nil
+}
+
+// registerIndexerFunc registers an indexer that is used to retrieve an event from the
+// underlying watch cache given an index.
+func (wci *watchCacheInterval) registerIndexerFunc(indexer func(int) *watchCacheEvent) {
+	wci.indexer = indexer
+}
+
+// Next returns the next item in the cache interval provided the cache
+// interval is still valid. An error is returned if the interval is
+// invalidated.
+func (wci *watchCacheInterval) Next() (*watchCacheEvent, bool, error) {
+	// if there are items in the buffer to return, return from
+	// the buffer.
+	event, exists := wci.intervalBuffer.next()
+	if exists {
+		return event, exists, nil
+	}
+	// if it is not possible to fill the interval
+	// buffer anymore, don't try and return.
+	if wci.startIndex >= wci.endIndex {
+		return nil, false, nil
+	}
+	wci.lock.Lock()
+	defer wci.lock.Unlock()
+
+	// check if the interval is invalidated.
+	if !wci.valid {
+		return nil, false, fmt.Errorf("cache interval invalidated")
+	}
+
+	wci.fillBuffer()
+	return wci.intervalBuffer.getNextEvent(), true, nil
+}
+
+func (wci *watchCacheInterval) fillBuffer() {
+	wci.intervalBuffer.wciBufferStart = 0
+	wci.intervalBuffer.wciBufferEnd = 0
+	for !wci.intervalBuffer.isFull() {
+		event := wci.indexer(wci.startIndex)
+		if event == nil {
+			break
+		}
+		wci.intervalBuffer.buffer[wci.intervalBuffer.wciBufferEnd] = event
+		wci.intervalBuffer.wciBufferEnd++
```

Comment:
wci.intervalBuffer.buffer[wci.intervalBuffer.wciBufferEnd++] = event

Reply:
Doing this isn't allowed in Go if I understand correctly since `++` ends up being a statement and not an expression.
Ref: https://golang.org/doc/faq#inc_dec 
Ex: https://play.golang.org/p/-uXOXAi54T1